embrace-android-core/src/main/kotlin/io/embrace/android/embracesdk/internal/comms/delivery/v2/SchedulingService.kt

```diff
@@ -0,0 +1,111 @@
+package io.embrace.android.embracesdk.internal.comms.delivery.v2
+
+import io.embrace.android.embracesdk.internal.clock.Clock
+import io.embrace.android.embracesdk.internal.comms.api.ApiResponse
+import io.embrace.android.embracesdk.internal.worker.BackgroundWorker
+import io.embrace.android.embracesdk.internal.worker.ScheduledWorker
+import java.io.File
+import java.util.Queue
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+
+interface SchedulingService {
+    fun checkPendingPayloads()
+}
+
+internal class SchedulingServiceImpl(
+    private val cacheService: CacheService2,
+    private val executionService: ExecutionService,
+    private val schedulingWorker: ScheduledWorker,
+    private val worker: BackgroundWorker,
+    private val clock: Clock
+) : SchedulingService {
+
+    private val retryQueue: Queue<File> = ConcurrentLinkedQueue()
+
+    override fun checkPendingPayloads() {
+        schedulingWorker.submit {
+            var ready = readyToSend()
+            while (ready) {
+                runCatching {
+                    getNextReadyPayload()?.let { payload ->
+                        val response = send(payload).get()
+                        if (response is ApiResponse.TooManyRequests) {
+                            blockEndpoint(response)
+                        }
+                    }
+                }
+
+                ready = readyToSend()
+            }
+            scheduleNextRetry()
+        }
+    }
+
+    private fun getNextReadyPayload(): File? {
```

Comment:
Should this be the priority of the scheduling service or should it be in the cache service?

Reply:
This service needs to be aware of more than just priority, but also if a particular endpoint is being 429'd or if the highest priority payload needs to be retried.

So if we deferred this to the CacheService, we'd have to make multiple reads disk. 

Now, I do think it's valuable to get a list of Files in sorted order by priority. I'll modify the interface to specify that in the comments and method name.